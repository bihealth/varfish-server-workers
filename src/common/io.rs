//! Common, IO-related code.

use std::{
    fs::File,
    io::{BufRead, BufReader, BufWriter, Write},
    path::Path,
};

use flate2::{bufread::MultiGzDecoder, write::GzEncoder, Compression};

/// Transparently open a file with gzip decoder.
pub fn open_read_maybe_gz<P>(path: P) -> Result<Box<dyn BufRead>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for reading", path.as_ref());
        let file = File::open(path)?;
        let bufreader = BufReader::new(file);
        let decoder = MultiGzDecoder::new(bufreader);
        Ok(Box::new(BufReader::new(decoder)))
    } else {
        tracing::trace!("Opening {:?} as plain text for reading", path.as_ref());
        let file = File::open(path).map(BufReader::new)?;
        Ok(Box::new(BufReader::new(file)))
    }
}

/// Transparently open afile with gzip encoder.
pub fn open_write_maybe_gz<P>(path: P) -> Result<Box<dyn Write>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for writing", path.as_ref());
        let file = File::create(path)?;
        let bufwriter = BufWriter::new(file);
        let encoder = GzEncoder::new(bufwriter, Compression::default());
        Ok(Box::new(encoder))
    } else {
        tracing::trace!("Opening {:?} as plain text for writing", path.as_ref());
        let file = File::create(path)?;
        Ok(Box::new(file))
    }
}

// The output is wrapped in a Result to allow matching on errors
// Returns an Iterator to the Reader of the lines of the file.
pub fn read_lines<P: AsRef<Path>>(
    filename: P,
) -> std::io::Result<std::io::Lines<std::io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(std::io::BufReader::new(file).lines())
}

#[cfg(test)]
mod test {
    use std::io::Read;

    #[rstest::rstest]
    #[case(true)]
    #[case(false)]
    fn open_write_maybe_gz(#[case] is_gzip: bool) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{:?}", is_gzip);

        let filename = if is_gzip { "test.txt" } else { "test.txt.gz" };
        let tmp_dir = temp_testdir::TempDir::default();

        {
            let mut f = super::open_write_maybe_gz(tmp_dir.join(filename))?;
            f.flush()?;
        }

        let mut f = std::fs::File::open(tmp_dir.join(filename)).map(std::io::BufReader::new)?;
        let mut buf = Vec::new();
        f.read_to_end(&mut buf)?;

        insta::assert_snapshot!(format!("{:x?}", &buf));

        Ok(())
    }

    #[test]
    fn read_lines() -> Result<(), anyhow::Error> {
        let lines = super::read_lines("tests/common/lines.txt")?.collect::<Result<Vec<_>, _>>()?;

        insta::assert_yaml_snapshot!(lines);

        Ok(())
    }
}
