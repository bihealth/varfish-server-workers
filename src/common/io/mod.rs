//! Common, IO-related code.

pub mod std;
