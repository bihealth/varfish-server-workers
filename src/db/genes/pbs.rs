//! Data structures for (de-)serialization as generated by `prost-build`.

include!(concat!(env!("OUT_DIR"), "/varfish.v1.gene.rs"));
