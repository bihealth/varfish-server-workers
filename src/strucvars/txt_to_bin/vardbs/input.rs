//! Code supporting the I/O of public database records and a common
//! `InputRecord` for common representation.

use serde::Deserialize;
use tracing::error;

use crate::strucvars::aggregate::output::Record as InhouseDbRecord;
use crate::strucvars::query::schema::SvType;

/// dbVar database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DbVarRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// number of overall carriers
    pub num_carriers: u32,
    /// type of the SV
    pub sv_type: String,
}

/// DGV database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DgvRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural variant type
    sv_type: String,
    /// Number of observed gains.
    observed_gains: u32,
    /// Number of observed losses
    observed_losses: u32,
}

/// DGV gold standard database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DgvGsRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin_outer: i32,
    /// outer end position, 0-based
    pub end_outer: i32,
    /// The structural variant type
    pub sv_sub_type: String,
    /// Number of carriers.
    pub num_carriers: u32,
}

/// ExAC CNV database record as read from TSV file for deserialization
/// from TSV.
#[derive(Deserialize, Debug)]
pub struct ExacRecord {
    /// chromosome name
    pub chromosome: String,
    /// outer start position, 0-based
    pub begin: i32,
    /// outer end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub sv_type: String,
}

/// Thousand Genomes SV database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct G1kRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub sv_type: String,
    /// Number of hom. alt. alleles.
    pub n_homalt: u32,
    /// Number of het. alt. alleles.
    pub n_het: u32,
}

/// gnomAD SV v2 database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadSv2Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of homozygous alternative carriers
    pub n_homalt: u32,
    /// Number of heterozygous carriers
    pub n_het: u32,
}

/// gnomAD SV v4 database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadSv4Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of male homozygous reference allele carriers.
    pub male_n_homref: u32,
    /// Number of male heterozygous alternate allele carriers.
    pub male_n_het: u32,
    /// Number of male homozygous alternate allele carriers.
    pub male_n_homalt: u32,
    /// Number of male hemizygous alternate allele carriers.
    pub male_n_hemiref: u32,
    /// Number of male hemizygous reference allele carriers.
    pub male_n_hemialt: u32,
    /// Number of female homozygous reference allele carriers.
    pub female_n_homref: u32,
    /// Number of female heterozygous alternate allele carriers.
    pub female_n_het: u32,
    /// Number of female homozygous alternate allele carriers.
    pub female_n_homalt: u32,
    /// Number of samples at this site (CNV only).
    pub cnv_n_total: u32,
    /// Number of samples with a CNV at this site (CNV only).
    pub cnv_n_var: u32,
}

/// gnomAD CNV v$ database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadCnv4Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of samples at this site (passing QC).
    pub n_total: u32,
    /// Number of samples with a CNV at this site (passing QC).
    pub n_var: u32,
}

/// Common type to convert input data to.
pub struct InputRecord {
    /// Chromosome of start position.
    pub chromosome: String,
    /// Chromosome of end position.
    pub chromosome2: String,
    /// SV type
    pub sv_type: SvType,
    /// 0-based begin position
    pub begin: i32,
    /// 0-based end position
    pub end: i32,
    /// Number of carriers (or alleles), depending on database.
    pub count: u32,
}

impl TryInto<Option<InputRecord>> for InhouseDbRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome,
            chromosome2: self.chromosome2,
            sv_type: self.sv_type,
            begin: self.begin,
            end: self.end,
            count: self.carriers,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DbVarRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.split(';').next().unwrap() {
            "alu_insertion"
            | "herv_insertion"
            | "insertion"
            | "line1_insertion"
            | "mobile_element_insertion"
            | "novel_sequence_insertion"
            | "sva_insertion" => SvType::Ins,
            "copy_number_gain" | "duplication" | "tandem_duplication" => SvType::Dup,
            "alu_deletion" | "copy_number_loss" | "deletion" | "herv_deletion"
            | "line1_deletion" | "sva_deletion" => SvType::Del,
            "copy_number_variation" => SvType::Cnv,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: 1,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DgvRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_ref() {
            "alu deletion"
            | "deletion"
            | "herv deletion"
            | "line1 deletion"
            | "mobile element deletion"
            | "loss"
            | "sva deletion" => SvType::Del,
            "alu insertion"
            | "herv insertion"
            | "insertion"
            | "line1 insertion"
            | "mobile element insertion"
            | "novel sequence insertion"
            | "sva insertion" => SvType::Ins,
            "duplication" | "gain" | "tandem duplication" => SvType::Dup,
            "sequence alteration" | "complex" => return Ok(None), // skip
            "gain+loss" | "CNV" => SvType::Cnv,
            "inversion" => SvType::Inv,
            "OTHER" => return Ok(None), // skip
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: self.observed_gains + self.observed_losses,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DgvGsRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_sub_type.as_ref() {
            "Gain" => SvType::Dup,
            "Loss" => SvType::Del,
            _ => {
                error!("sv_type = {}", &self.sv_sub_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin_outer,
            end: self.end_outer,
            sv_type,
            count: self.num_carriers,
        }))
    }
}

impl TryInto<Option<InputRecord>> for ExacRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_ref() {
            "DUP" => SvType::Dup,
            "DEL" => SvType::Del,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: 1,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadSv2Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.svtype.as_str() {
            "CPX" => return Ok(None), // no correspondence
            "CTX" | "BND" => SvType::Bnd,
            "DEL" => SvType::Del,
            "DUP" => SvType::Dup,
            "INS" => SvType::Ins,
            "INV" => SvType::Inv,
            "MCNV" => SvType::Cnv,
            _ => {
                error!("sv_type = {}", &self.svtype);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin - 1,
            end: self.end,
            sv_type,
            count: self.n_homalt + self.n_het,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadCnv4Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type: match self.svtype.as_str() {
                "DEL" => SvType::Del,
                "DUP" => SvType::Dup,
                _ => {
                    error!("sv_type = {}", &self.svtype);
                    return Err("unknown SV type");
                }
            },
            count: self.n_var,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadSv4Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type: match self.svtype.as_str() {
                "BND" => SvType::Bnd,
                "CNV" => SvType::Cnv,
                "DEL" => SvType::Del,
                "DUP" => SvType::Dup,
                "INS" => SvType::Ins,
                "INV" => SvType::Inv,
                _ => {
                    error!("sv_type = {}", &self.svtype);
                    return Err("unknown SV type");
                }
            },
            count: self.male_n_het
                + self.male_n_homalt
                + self.male_n_hemialt
                + self.female_n_het
                + self.female_n_homalt
                + self.cnv_n_var,
        }))
    }
}

impl TryInto<Option<InputRecord>> for G1kRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_str() {
            "CN0" | "CNV" => SvType::Cnv,
            "DEL" => SvType::Del,
            "DEL_ALU" | "DEL_HERV" | "DEL_LINE1" | "DEL_SVA" => SvType::Del,
            "DUP" => SvType::Dup,
            "INV" => SvType::Inv,
            "INS" | "INS:ME:ALU" | "INS:ME:LINE1" | "INS:ME:SVA" => SvType::Ins,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: self.n_homalt + self.n_het,
        }))
    }
}
