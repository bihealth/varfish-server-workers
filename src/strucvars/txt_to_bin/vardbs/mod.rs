//! Code for converting other structural variant database to binary (incl. in-house).

use std::fs::File;
use std::io::Write;
use std::path::Path;
use std::time::Instant;

use anyhow::anyhow;
use prost::Message;
use thousands::Separable;

use crate::common::{build_chrom_map, trace_rss_now};
use crate::pbs::svs::{BackgroundDatabase, BgDbRecord};
use crate::strucvars::aggregate::output::Record as InhouseDbRecord;
use crate::strucvars::query::schema::SvType;

use self::input::InputRecord;

mod input;

/// Known input file types.
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum InputFileType {
    Dbvar,
    Dgv,
    DgvGs,
    Exac,
    G1k,
    GnomadSv2,
    GnomadCnv4,
    GnomadSv4,
    InhouseDb,
}
/// Deserialize from CSV reader to an `Option<records::InputRecord>`
fn deserialize_loop<Rec>(
    reader: &mut csv::Reader<Box<dyn std::io::BufRead>>,
) -> Result<Vec<BgDbRecord>, anyhow::Error>
where
    Rec: core::fmt::Debug + TryInto<Option<InputRecord>> + for<'de> serde::Deserialize<'de>,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: core::fmt::Debug,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: Send,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: std::marker::Sync,
{
    let chrom_map = build_chrom_map();
    let mut result = Vec::new();

    for record in reader.deserialize() {
        let record: Rec = record?;
        let maybe_record: Option<InputRecord> = record
            .try_into()
            .map_err(|err| anyhow!("problem with parsing: {:?}", &err))?;
        if let Some(record) = maybe_record {
            result.push(BgDbRecord {
                chrom_no: *chrom_map
                    .get(&record.chromosome)
                    .unwrap_or_else(|| panic!("unknown chrom: {:?}", &record.chromosome))
                    as i32,
                chrom_no2: *chrom_map
                    .get(&record.chromosome2)
                    .unwrap_or_else(|| panic!("unknown chrom2: {:?}", &record.chromosome2))
                    as i32,
                sv_type: match record.sv_type {
                    SvType::Del => crate::pbs::svs::SvType::Del,
                    SvType::Dup => crate::pbs::svs::SvType::Dup,
                    SvType::Inv => crate::pbs::svs::SvType::Inv,
                    SvType::Ins => crate::pbs::svs::SvType::Ins,
                    SvType::Bnd => crate::pbs::svs::SvType::Bnd,
                    SvType::Cnv => crate::pbs::svs::SvType::Cnv,
                } as i32,
                start: record.begin + 1,
                stop: record.end,
                count: record.count,
            });
        }
    }

    Ok(result)
}

/// Branch around `deserialize_loop`.
pub fn deserialize_branch(
    input_type: InputFileType,
    reader: &mut csv::Reader<Box<dyn std::io::BufRead>>,
) -> Result<Vec<BgDbRecord>, anyhow::Error> {
    match input_type {
        InputFileType::Dbvar => deserialize_loop::<input::DbVarRecord>(reader),
        InputFileType::Dgv => deserialize_loop::<input::DgvRecord>(reader),
        InputFileType::DgvGs => deserialize_loop::<input::DgvGsRecord>(reader),
        InputFileType::Exac => deserialize_loop::<input::ExacRecord>(reader),
        InputFileType::G1k => deserialize_loop::<input::G1kRecord>(reader),
        InputFileType::InhouseDb => deserialize_loop::<InhouseDbRecord>(reader),
        InputFileType::GnomadSv2 => deserialize_loop::<input::GnomadSv2Record>(reader),
        InputFileType::GnomadCnv4 => deserialize_loop::<input::GnomadCnv4Record>(reader),
        InputFileType::GnomadSv4 => deserialize_loop::<input::GnomadSv4Record>(reader),
    }
}

/// Perform conversion to protobuf `.bin` file.
pub fn convert_to_bin<P, Q>(
    path_input_tsv: P,
    path_output: Q,
    input_type: InputFileType,
) -> Result<(), anyhow::Error>
where
    P: AsRef<Path>,
    Q: AsRef<Path>,
{
    // Setup CSV reader for BED file - header is written as comment and must be
    // ignored.
    let mut reader = csv::ReaderBuilder::new()
        .has_headers(false)
        .comment(Some(b'#'))
        .delimiter(b'\t')
        .from_reader(mehari::common::io::std::open_read_maybe_gz(
            path_input_tsv.as_ref(),
        )?);
    let before_parsing = Instant::now();

    let records = deserialize_branch(input_type, &mut reader)?;
    let bg_db = BackgroundDatabase { records };

    tracing::debug!(
        "total time spent reading {} records: {:?}",
        bg_db.records.len().separate_with_commas(),
        before_parsing.elapsed()
    );
    trace_rss_now();

    let before_writing = Instant::now();
    let mut output_file = File::create(&path_output)?;
    output_file.write_all(&bg_db.encode_to_vec())?;
    output_file.sync_all()?;
    tracing::debug!(
        "total time spent writing {} records: {:?}",
        bg_db.records.len().separate_with_commas(),
        before_writing.elapsed()
    );

    Ok(())
}

#[cfg(test)]
mod test {
    use super::InputFileType;

    #[rstest::rstest]
    #[case::dbvar(
        InputFileType::Dbvar,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dbvar.bed.gz"
    )]
    #[case::dgv(
        InputFileType::Dgv,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dgv.bed.gz"
    )]
    #[case::dgv_gs(
        InputFileType::DgvGs,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dgv_gs.bed.gz"
    )]
    #[case::exac(
        InputFileType::Exac,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/exac.bed.gz"
    )]
    #[case::g1k(
        InputFileType::G1k,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/g1k.bed.gz"
    )]
    #[case::gnomad_sv2(
        InputFileType::GnomadSv2,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/gnomad_sv.bed.gz"
    )]
    #[case::gnomad_cnv4(
        InputFileType::GnomadCnv4,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch38/strucvar/gnomad-cnv.bed.gz"
    )]
    #[case::gnomad_sv4(
        InputFileType::GnomadSv4,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch38/strucvar/gnomad-sv.bed.gz"
    )]
    #[case::inhouse_db(
        InputFileType::InhouseDb,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/inhouse.tsv"
    )]
    fn test_deserialize_branch(
        #[case] input_type: InputFileType,
        #[case] path_input: &str,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!(
            "{:?}-{}",
            input_type,
            path_input
                .split('/')
                .last()
                .unwrap()
                .split('.')
                .next()
                .unwrap()
        );

        let mut reader = csv::ReaderBuilder::new()
            .has_headers(false)
            .comment(Some(b'#'))
            .delimiter(b'\t')
            .from_reader(mehari::common::io::std::open_read_maybe_gz(path_input)?);

        let records = super::deserialize_branch(input_type, &mut reader)?;
        insta::assert_yaml_snapshot!(records);

        Ok(())
    }
}
