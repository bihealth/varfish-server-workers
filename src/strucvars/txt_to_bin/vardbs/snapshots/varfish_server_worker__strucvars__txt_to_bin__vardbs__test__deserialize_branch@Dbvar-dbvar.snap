---
source: src/strucvars/txt_to_bin/vardbs/mod.rs
expression: records
---
- svType: SV_TYPE_DUP
  start: 1
  stop: 2300000
  count: 1
- svType: SV_TYPE_DUP
  start: 1
  stop: 16200000
  count: 1
- svType: SV_TYPE_DUP
  start: 10001
  stop: 19818
  count: 1
- svType: SV_TYPE_DUP
  start: 10001
  stop: 22118
  count: 1
- svType: SV_TYPE_DUP
  start: 10001
  stop: 52000
  count: 1
- svType: SV_TYPE_DUP
  start: 10001
  stop: 82189
  count: 1
- svType: SV_TYPE_DUP
  start: 10001
  stop: 87466
  count: 1
- svType: SV_TYPE_DUP
  start: 10001
  stop: 88143
  count: 1
- svType: SV_TYPE_DUP
  start: 10001
  stop: 127330
  count: 1
- svType: SV_TYPE_DUP
  start: 10001
  stop: 356281
  count: 1
- svType: SV_TYPE_DUP
  start: 10001
  stop: 717485
  count: 1
- svType: SV_TYPE_INS
  start: 10059
  stop: 10059
  count: 1
- svType: SV_TYPE_DUP
  start: 10351
  stop: 13950
  count: 1
- svType: SV_TYPE_DUP
  start: 10351
  stop: 18900
  count: 1
- svType: SV_TYPE_DUP
  start: 10377
  stop: 177417
  count: 1
- svType: SV_TYPE_DUP
  start: 10377
  stop: 471368
  count: 1
- svType: SV_TYPE_DUP
  start: 10377
  stop: 812484
  count: 1
- svType: SV_TYPE_DUP
  start: 10377
  stop: 998287
  count: 1
- svType: SV_TYPE_INS
  start: 10380
  stop: 10380
  count: 1
- svType: SV_TYPE_DUP
  start: 10469
  stop: 23975
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 79204
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 79589
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 79864
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 80652
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 81231
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 81298
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 81448
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 81829
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 86730
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 87005
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 87096
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 87610
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 87845
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 89163
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 90440
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 91558
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 91591
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 92013
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 92392
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 177368
  count: 1
- svType: SV_TYPE_DUP
  start: 10499
  stop: 235938
  count: 1
- svType: SV_TYPE_DUP
  start: 10501
  stop: 78200
  count: 1
- svType: SV_TYPE_DUP
  start: 10529
  stop: 177368
  count: 1
- svType: SV_TYPE_INS
  start: 10627
  stop: 10627
  count: 1
- svType: SV_TYPE_INS
  start: 10657
  stop: 10657
  count: 1
- svType: SV_TYPE_INS
  start: 10726
  stop: 10726
  count: 1
- svType: SV_TYPE_INS
  start: 10752
  stop: 10752
  count: 1
- svType: SV_TYPE_DUP
  start: 10801
  stop: 14400
  count: 1
- svType: SV_TYPE_DUP
  start: 10801
  stop: 114750
  count: 1
- svType: SV_TYPE_DUP
  start: 11099
  stop: 47000
  count: 1
- svType: SV_TYPE_DUP
  start: 11100
  stop: 29200
  count: 1
- svType: SV_TYPE_DUP
  start: 11189
  stop: 36287
  count: 1
- svType: SV_TYPE_DUP
  start: 11189
  stop: 36787
  count: 1
- svType: SV_TYPE_CNV
  start: 11481
  stop: 35484
  count: 1
- svType: SV_TYPE_DUP
  start: 11599
  stop: 72400
  count: 1
- svType: SV_TYPE_DUP
  start: 11600
  stop: 26700
  count: 1
- svType: SV_TYPE_CNV
  start: 11601
  stop: 15700
  count: 1
- svType: SV_TYPE_CNV
  start: 11601
  stop: 20700
  count: 1
- svType: SV_TYPE_CNV
  start: 12048
  stop: 89237
  count: 1
- svType: SV_TYPE_CNV
  start: 12048
  stop: 177033
  count: 1
- svType: SV_TYPE_DUP
  start: 12601
  stop: 27100
  count: 1
- svType: SV_TYPE_CNV
  start: 12701
  stop: 13400
  count: 1
- svType: SV_TYPE_DUP
  start: 12701
  stop: 17551
  count: 1
- svType: SV_TYPE_DUP
  start: 12751
  stop: 17501
  count: 1
- svType: SV_TYPE_DUP
  start: 12751
  stop: 17551
  count: 1
- svType: SV_TYPE_DUP
  start: 12751
  stop: 17751
  count: 1
- svType: SV_TYPE_DUP
  start: 12801
  stop: 17501
  count: 1
- svType: SV_TYPE_CNV
  start: 12801
  stop: 17551
  count: 1
- svType: SV_TYPE_DUP
  start: 13516
  stop: 91073
  count: 1
- svType: SV_TYPE_DUP
  start: 14701
  stop: 17501
  count: 1
- svType: SV_TYPE_DUP
  start: 14801
  stop: 17501
  count: 1
- svType: SV_TYPE_DUP
  start: 14874
  stop: 368661
  count: 1
- svType: SV_TYPE_CNV
  start: 15801
  stop: 16100
  count: 1
- svType: SV_TYPE_CNV
  start: 16201
  stop: 20700
  count: 1
- svType: SV_TYPE_DUP
  start: 18888
  stop: 35451
  count: 1
- svType: SV_TYPE_DUP
  start: 19225
  stop: 4401691
  count: 1
- svType: SV_TYPE_INS
  start: 19742
  stop: 28927
  count: 1
- svType: SV_TYPE_CNV
  start: 19851
  stop: 20551
  count: 1
- svType: SV_TYPE_DUP
  start: 20141
  stop: 242707
  count: 1
- svType: SV_TYPE_DUP
  start: 20141
  stop: 356281
  count: 1
- svType: SV_TYPE_DUP
  start: 20141
  stop: 606093
  count: 1
- svType: SV_TYPE_DUP
  start: 20141
  stop: 887162
  count: 1
- svType: SV_TYPE_CNV
  start: 20701
  stop: 95200
  count: 1
- svType: SV_TYPE_INS
  start: 20848
  stop: 20848
  count: 1
- svType: SV_TYPE_INS
  start: 20853
  stop: 20853
  count: 1
- svType: SV_TYPE_DUP
  start: 21064
  stop: 28577
  count: 1
- svType: SV_TYPE_DUP
  start: 21064
  stop: 35983
  count: 1
- svType: SV_TYPE_DUP
  start: 21411
  stop: 82189
  count: 1
- svType: SV_TYPE_DUP
  start: 21829
  stop: 88143
  count: 1
- svType: SV_TYPE_DUP
  start: 22118
  stop: 22723
  count: 1
- svType: SV_TYPE_DUP
  start: 22118
  stop: 35020
  count: 1
- svType: SV_TYPE_DUP
  start: 22723
  stop: 23661
  count: 1
- svType: SV_TYPE_CNV
  start: 22801
  stop: 26900
  count: 1
- svType: SV_TYPE_DUP
  start: 23661
  stop: 24819
  count: 1
- svType: SV_TYPE_DUP
  start: 23946
  stop: 88271
  count: 1
- svType: SV_TYPE_CNV
  start: 24301
  stop: 25700
  count: 1
- svType: SV_TYPE_CNV
  start: 24301
  stop: 26900
  count: 1
- svType: SV_TYPE_CNV
  start: 24301
  stop: 128700
  count: 1
- svType: SV_TYPE_DUP
  start: 24310
  stop: 35804991
  count: 1
